text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,651 @@
+     `assemblies/${assembly.name}/versions.json` object, which contains a list of
+     all versions ever indexed for the package, matched to the versions' current
+     status (`deprecated`, `latest`, ...):
+
+     Name        | Description
+     ------------|----------------------------------------------------------------
+     `v${major}` | All versions part of this major version stream, and their status
+
+     [SemVer]: https://semver.org/spec/v2.0.0.html
+
+     In order to ensure consistency of the contents of the `versions.json`
+     object, its content will always be generated from the live inventory of
+     objects in S3 (and not incrementally using the update trigger event), and
+     the last event timestamp will be written in an object metadata attribute.
+     The S3 Bucket will have object versioning enabled, and once a new version
+     of the `versions.json` object has been created, its metadata will be
+     checked against any previous versions of the object, and the "current"
+     version will be set to the one with the latest event timestamp if needed.
+
+   The operator of private instances can provide an optional SNS topic which
+   will be notified once the information for a new package version is ready to
+   be browsed.
+
+1. **Doc-Gen:** A series of Lambda functions prepare language-specific assembly
+   files for each configured language, with adjusted naming conventions, and
+   updated sample code fragments, and proceeds to store those at:
+   `assemblies/${assembly.name}/v${assembly.version}/assembly-${lang}.json`.
+   This transformation is backed by the `jsii-rosetta` tool, which is part of
+   the [jsii project][jsii].
+
+   [jsii]: https://aws.github.io/jsii
+
+   - Alternatively, this work wan be pushed up-front to package build time, by
+     baking this feature into the `jsii` compiler directly. This would have the
+     benefits of simplifying the processing, as all necessary dependencies are
+     necessarily present during `jsii` compilation phase.
+
+1. **Latest Updater:** A Lambda function keeps the `latest.json` object updated
+   with the last 20 package versions indexed (according to the `time` field
+   reported by the ingestion component). S3 Object Versioning will be used to
+   ensure concurrent updates to the content result in a consistent end state,
+   by listing object versions before and after updating the `latest.json` object
+   and performing the necessary corrections if needed.
+
+1. **Catalog Builder:** A Lambda function keeps the `catalog.json` object
+   updated with the latest versions of each package's major version lines, which
+   backs the websites' search page. The object may be sharded if it becomes too
+   large. S3 Object Versioning will be used to ensure concurrent updates to the
+   content result in a consistent end state, by listing object versions before
+   and after updating the `catalog.json` object and performing the necessary
+   corrections if needed.
+
```

Comment:
Please include the public URL schema here (basically that's our backend API).

If I recall we discussed that the backend URLs will generally derive from the frontend URLs with an `api.` subdomain or something like that:

Get the JSII manifest of the latest version of a module:

```
GET https://api.constructs.dev/packages/packages[/@<scope>]/<name>
```

Get the JSII manifest of a specific version:

```
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>
```

Go do languages come into play here?

```
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/python
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/java
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/go
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/typescript
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/javascript
GET https://api.constructs.dev/packages/@<scope>]/<name>/v/<version>/csharp
```

(not 100% sure, but something along those lines).